File: apps/web/src/lib/validations/patient.ts
import { z } from "zod";

/** Zod schema for patient registration form validation. */
export const patientCreateSchema = z.object({
  first_name: z.string().min(1).max(100),
  middle_name: z.string().max(100).nullable().optional(),
  last_name: z.string().min(1).max(100),
  date_of_birth: z.string().min(1),
  gender: z.enum(["male", "female", "other"]),
  national_id: z.string().max(50).nullable().optional(),
  passport_number: z.string().max(50).nullable().optional(),
  phone_number: z.string().min(9).max(20),
  alternate_phone: z.string().max(20).nullable().optional(),
  email: z.string().email().max(255).nullable().optional().or(z.literal("")),
  county: z.string().max(100).nullable().optional(),
  sub_county: z.string().max(100).nullable().optional(),
  ward: z.string().max(100).nullable().optional(),
  village: z.string().max(200).nullable().optional(),
  postal_address: z.string().max(200).nullable().optional(),
  occupation: z.string().max(100).nullable().optional(),
  marital_status: z.enum(["single", "married", "divorced", "widowed"]).nullable().optional(),
  next_of_kin_name: z.string().max(200).nullable().optional(),
  next_of_kin_phone: z.string().max(20).nullable().optional(),
  next_of_kin_relationship: z.string().max(50).nullable().optional(),
  insurance_provider: z.string().max(100).nullable().optional(),
  insurance_member_number: z.string().max(100).nullable().optional(),
  sha_number: z.string().max(50).nullable().optional(),
  blood_group: z.string().regex(/^(A|B|AB|O)[+-]$/).nullable().optional().or(z.literal("")),
  allergies: z.array(z.string()).nullable().optional(),
  chronic_conditions: z.array(z.string()).nullable().optional(),
});

export type PatientCreateFormData = z.infer<typeof patientCreateSchema>;

File: services/api-gateway/app/models/patient.py
import uuid
from datetime import date, datetime

from sqlalchemy import Date, DateTime, Index, String, Text, func
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column

from app.database import Base
from app.models.base import AuditMixin


class Patient(AuditMixin, Base):
    """
    Patient demographics and registration data.
    Multi-tenant: every query filtered by facility_id from JWT.
    """

    __tablename__ = "patients"
    __table_args__ = (
        Index("ix_patients_facility_national_id", "facility_id", "national_id"),
        Index("ix_patients_facility_phone", "facility_id", "phone_number"),
        Index(
            "ix_patients_facility_name",
            "facility_id",
            "first_name",
            "last_name",
        ),
        Index("ix_patients_mrn", "facility_id", "mrn", unique=True),
    )

    # MRN (Medical Record Number) — facility-specific
    mrn: Mapped[str] = mapped_column(String(50), nullable=False)

    # Demographics
    first_name: Mapped[str] = mapped_column(String(100), nullable=False)
    middle_name: Mapped[str | None] = mapped_column(String(100))
    last_name: Mapped[str] = mapped_column(String(100), nullable=False)
    date_of_birth: Mapped[date] = mapped_column(Date, nullable=False)
    gender: Mapped[str] = mapped_column(String(20), nullable=False)

    # Identification
    national_id: Mapped[str | None] = mapped_column(String(50))
    passport_number: Mapped[str | None] = mapped_column(String(50))

    # Contact
    phone_number: Mapped[str] = mapped_column(String(20), nullable=False)
    alternate_phone: Mapped[str | None] = mapped_column(String(20))
    email: Mapped[str | None] = mapped_column(String(255))

    # Address (Kenya-specific)
    county: Mapped[str | None] = mapped_column(String(100))
    sub_county: Mapped[str | None] = mapped_column(String(100))
    ward: Mapped[str | None] = mapped_column(String(100))
    village: Mapped[str | None] = mapped_column(String(200))
    postal_address: Mapped[str | None] = mapped_column(String(200))

    # Personal
    occupation: Mapped[str | None] = mapped_column(String(100))
    marital_status: Mapped[str | None] = mapped_column(String(20))

    # Next of Kin
    next_of_kin_name: Mapped[str | None] = mapped_column(String(200))
    next_of_kin_phone: Mapped[str | None] = mapped_column(String(20))
    next_of_kin_relationship: Mapped[str | None] = mapped_column(String(50))

    # Insurance
    insurance_provider: Mapped[str | None] = mapped_column(String(100))
    insurance_member_number: Mapped[str | None] = mapped_column(String(100))
    sha_number: Mapped[str | None] = mapped_column(String(50))

    # Medical
    blood_group: Mapped[str | None] = mapped_column(String(10))
    allergies: Mapped[list | None] = mapped_column(JSONB)  # type: ignore[type-arg]
    chronic_conditions: Mapped[list | None] = mapped_column(JSONB)  # type: ignore[type-arg]

    # Photo
    photo_url: Mapped[str | None] = mapped_column(String(500))

    # FHIR
    fhir_id: Mapped[str | None] = mapped_column(String(100))
